File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
import re

def solution(dartResult):
    bonus = {"S" : 1, "D" : 2, "T" : 3}
    option = {"" : 1, "*" : 2, "#" : -1}
    
    p = re.compile("(\d+)([SDT])([*#])?")
    result = p.findall(dartResult)
    
    answer = [0, 0, 0]
    for i in range(3):
        if i > 0 and result[i][2] == '*':
            answer[i - 1] *= 2
        answer[i] = int(result[i][0]) ** bonus[result[i][1]] * option[result[i][2]]
    return sum(answer)

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;
import java.util.regex.*;

class Solution {
    public int solution(String dartResult) {
        Map<String, Integer> map = new HashMap<>();
        map.put("S", 1); map.put("D", 2); map.put("T", 3);
        map.put("*", 2); map.put("#", -1);
        
        Pattern pattern = Pattern.compile("(\\d+)([SDT])([*#])?");
        Matcher matcher = pattern.matcher(dartResult);
        
        int idx = 0;
        String score, bonus, option;
        int[] result = new int[3];
        while (matcher.find()) {
            score = matcher.group(1); bonus = matcher.group(2); option = matcher.group(3);
            
            if (idx > 0 && option != null && option.equals("*")) {
                result[idx - 1] *= 2;
            }
 
            result[idx] = (int) Math.pow(Integer.parseInt(score), map.get(bonus));
            if (option != null) {
                result[idx] *= map.get(option);
            }
            idx += 1;
        }
        
        int answer = 0;
        for (int i = 0; i < 3; i++) {
            answer += result[i];
        }
        return answer;
    }
}